README.md
# dora-plugin-weinre

[![NPM version](https://img.shields.io/npm/v/dora-plugin-weinre.svg?style=flat)](https://npmjs.org/package/dora-plugin-weinre)
[![Build Status](https://img.shields.io/travis/dora-js/dora-plugin-weinre.svg?style=flat)](https://travis-ci.org/dora-js/dora-plugin-weinre)
[![Coverage Status](https://img.shields.io/coveralls/dora-js/dora-plugin-weinre.svg?style=flat)](https://coveralls.io/r/dora-js/dora-plugin-weinre)
[![NPM downloads](http://img.shields.io/npm/dm/dora-plugin-weinre.svg?style=flat)](https://npmjs.org/package/dora-plugin-weinre)

dora plugin for weinre.

---

## Usage

```bash
$ npm i dora dora-plugin-weinre -SD
$ ./node_modules/.bin/dora --plugins 'weinre?httpPort=8888'
```

## Param

```javascript
defaultOptions = {
  httpPort: 8990,
  boundHost: 'yourIP',
  verbose: false,
  debug: false,
  readTimeout: 5,
  deathTimeout: 15,
  help: false,
};
```

## Test

```bash
$ npm test
```

## LICENSE

MIT


HISTORY.md
# History

---

## 0.3.1

- add npm run compile

## 0.3.0

- enhance inject weinre https://github.com/dora-js/dora-plugin-weinre/pull/1

## 0.2.0

- support dora@0.3

## 0.1.1

- fix: babel-polyfill should be dependencies 

## 0.1.0

- initial version


test/index-test.js
import dora from 'dora';
import { join } from 'path';
import request from 'supertest';

const localIP = require('internal-ip')();
const port = '1234';

describe('index', () => {
  describe('weinre', () => {
    const cwd = process.cwd();
    before(done => {
      process.chdir(join(__dirname, './fixtures/normal'));
      dora({
        port,
        plugins: ['dora-plugin-webpack', '../../../src/index?{httpPort:8888}'],
        cwd: join(__dirname, './fixtures/normal'),
      }, done);
    });

    after(() => {
      process.chdir(cwd);
    });

    it('GET weinre /target/target-script-min.js#anonymous', done => {
      request(`http://${localIP}:8888`)
        .get('/target/target-script-min.js#anonymous')
        .expect(200)
        .end((err, res) => {
          if (err) return done(err);
          if (res.text.indexOf("modjewel.require('weinre/target/Target').main()") < 0) {
            const e = new Error('/target/target-script-min.js#anonymous is not correct');

            return done(e);
          }

          return done();
        });
    });

    it('GET /index.html is injected the script /target/target-script-min.js#anonymous', done => {
      request(`http://localhost:${port}`)
        .get('/index.html')
        .expect(200)
        .end((err, res) => {
          if (err) return done(err);
          if (res.text.indexOf('// weinre') < 0) {
            const e = new Error('/target/target-script-min.js#anonymous is not injected');

            return done(e);
          }

          return done();
        });
    });

    it('GET /lackdoctype.html is injected the script livereload.js', done => {
      request(`http://localhost:${port}`)
        .get('/lackdoctype.html')
        .expect(200)
        .end((err, res) => {
          if (err) return done(err);
          if (res.text.indexOf('// weinre') < 0) {
            const e = new Error('/target/target-script-min.js#anonymous is not injected');

            return done(e);
          }

          return done();
        });
    });

    it('GET /index.js should be handled', done => {
      request(`http://localhost:${port}`)
        .get('/index.js')
        .expect(200)
        .end((err, res) => {
          if (err) return done(err);
          if (res.text.indexOf('// weinre') < 0) {
            const e = new Error('other types of files should be handled');

            return done(e);
          }

          return done();
        });
    });
  });
});


src/injectScript.js
import ConcatSource from 'webpack-core/lib/ConcatSource';
import { getInjectWeinreContent } from './util';

export default class InjectScript {
  static defaults = {
    httpPort: 8990,
    boundHost: 'localhost',
  };

  constructor(options) {
    this.options = { ...InjectScript.defaults, ...options };
  }

  apply(compiler) {
    compiler.plugin('compilation', compilation => {
      const opts = this.options;
      compilation.plugin('optimize-chunk-assets', (chunks, callback) => {
        chunks.forEach(chunk => {
          chunk.files.filter(file => /.(js)$/.test(file)).forEach(file => {
            const injectContent = getInjectWeinreContent(opts.injectHost, opts.port);
            compilation.assets[file] = new ConcatSource(
              injectContent,
              '\n',
              compilation.assets[file]
            );
          });
        });

        callback();
      });
    });
  }
}


src/index.js
import { existsSync, readFileSync } from 'fs';
import { parse } from 'url';
import { join } from 'path';

import { getInjectWeinreContent } from './util';
import InjectScript from './injectScript';

const localIP = require('internal-ip')();

/**
 * weinre code
 * https://git-wip-us.apache.org/repos/asf?p=cordova-weinre.git;a=blob;f=weinre.server/lib/cli.js;h=3c4b50784a538af7fdbb471d56f35e11802ef0f4;hb=cc97191c79ca260b5bbf83b19d5cfc3e02468cac
 */
require('coffee-script');
import { run } from 'weinre';

let defaultOpts = {
  httpPort: 8990,
  boundHost: localIP,
  verbose: false,
  debug: false,
  readTimeout: 5,
  deathTimeout: 15,
  help: false,
};

export default {
  name: 'dora-plugin-weinre',

  'middleware.before'() {
    const { log } = this;
    run(defaultOpts);
    log.info(`weinre is started, servering at http://${defaultOpts.boundHost}:${defaultOpts.httpPort}`);
  },

  'middleware'() {
    const { cwd, get } = this;

    const compiler = get('compiler');
    if (!compiler) {
      throw new Error('[error] must used together with dora-plugin-webpack');
    }
    return function* middleFunc(next) {
      const pathName = parse(this.url).pathname;
      const fileName = pathName === '/' ? 'index.html' : pathName;
      const filePath = join(cwd, fileName);
      const isHTML = /\.html?$/.test(fileName);
      if (isHTML && existsSync(filePath)) {
        const injectContent = getInjectWeinreContent(defaultOpts.boundHost, defaultOpts.httpPort);
        const injectScript = `<script>${injectContent}</script>`;
        let content = readFileSync(filePath, 'utf-8');
        content = injectScript + content;
        this.body = content;

        return;
      }
      yield next;
    };
  },

  'webpack.updateConfig.finally'(webpackConfig) {
    const { query } = this;

    defaultOpts = { ...defaultOpts, ...query };

    webpackConfig.plugins.push(new InjectScript({
      boundHost: defaultOpts.boundHost,
      httpPort: defaultOpts.httpPort,
    }));

    return webpackConfig;
  },
};
